# Problem solving and programming
## 12 june 2019

## string slicing

In [26]:
s1="python123"
#s1[-1]
s1[len(s1)-1]
s1[0:4]
s1[-1:-2]
s1[-2:]    #accessing the last two char in a string
s1[1:]     #accessing all characters after 2nd position
s1[1:-1]   #accessing all char except first and last
s1[len(s1)//2]   #accessing middle char of given string
s1[-1::-1]#reverse of string
s1[-1:-3:-1]#accessing last two char in reverse order
s1[len(s1)//2:]#reverse the middle two char in an even length string
s1[1:3]
s1[-1:]
s1[1::3]
s1[0::-1]



'p'

## leapyear

In [30]:
#check given number is leap year or not
def isleapyear(year):
    if year%400==0 or (year%4==0) and (year%100!=0):
       return "true"
    return "false"
isleapyear(2016)


'true'

In [32]:
#function to count the number of digits in a given number
def digit(d):
    k=[len(str(d))]
    return k
digit(123)

[3]

In [ ]:
#greatest of the numbers
def greatest(a):
    a=(12,14,15)
    

## iteration
### for
### while

In [36]:
#function to print n natural numbers

def printNnaturalnumbers(n):
    for counter in range(1,n+1):
        print(counter, end=" ")
    return
printNnaturalnumbers(20)

1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 

In [40]:
def printNnaturalnumbers(n):
    counter=1
    while counter <= n:
        print(counter, end=" ")
        counter=counter+1
    return
printNnaturalnumbers(10)

1 2 3 4 5 6 7 8 9 10 

In [5]:
#function to print all numbers divisible by 6
#and not a factor of 100 in a given range(lb, ub) inclusive

def printallnumbers(lb,ub):
    for i in range(lb,ub+1):
        if (i % 6==0) and (100 % i!=0):
           print(i, end=" ")
    return
printallnumbers(100,200)


102 108 114 120 126 132 138 144 150 156 162 168 174 180 186 192 198 

In [26]:
#avg of cubes of all even numbers in a given range
def printallnumbers(lb,ub):
    count=0
    sum=0
    for i in range(lb,ub+1):
        if i%2==0:
            #print(i)
            k=i**3
            #print(k)
            count=count+1
            sum=sum+k
            #print(k)
            #print(sum)
            #print(count)
    p=sum/count
    return p
printallnumbers(1,10)
        
        


360.0

In [32]:
#function to generate the list of factors of given number
def printfactors(n):
    for i in range(1,n+1):
        if n%i==0:
            print(i)
    return
printfactors(10)
        

1
2
5
10


In [15]:
#check given number is prime 
def prime(n):
    count=0
    for i in range(1,n+1):
        if
            count=count+1
    if count==2:
        print("prime")
    else:
        print("not prime")
    return
prime(3)
            
    

not prime


In [40]:
#average of first n prime numbers
def primeavg():
   
   l=int(input('enter 1st no  : '))
   u=int(input('enter 2nd no  : '))
   sum=0
   count=0
   for num in range(l,u + 1):
       if num > 1:
          for i in range(2,num):
             #print(i)
             if (num % i) == 0:
                  break
          else:
           #print(num)
           sum=sum+num
           count=count+1
   return (sum/count)
   
primeavg()

        
        

enter 1st no  : 10
enter 2nd no  : 20


15.0

In [38]:
#function to generate all perfect numbers in a given range

def factors_sum(i):
    sum=0
    for j in range(1,i):
        if i%j==0:
            sum+=j
    return sum
def perfect(lb,ub):
    sum=0
    for i in range(lb,ub):
        x=factors_sum(i)
        if i==x:
            print(i,end=" ")
    return
perfect(1,1500)

6 28 496 

In [17]:
#function to print the alternate values in a range
#[500,550]-->inclusive set i.e,starting value is 500 and ending value is 550
#(500,550)-->exclusive set i.e,starting value is 501 and ending value is 549
#range(500,550)-->500 501 502 ....549
#all set functions in python have start value inclusive 

def alternatevalues(start,end):
    for value in range(start,end+1,4):
        print (value, end=" ")
    return
alternatevalues(500,550)

500 504 508 512 516 520 524 528 532 536 540 544 548 

In [25]:
#function to print reverse of given range in the same line

def reverserange(start,end):
    for value in range(end,start-1,-1):
        print(value, end=" ")
    return
reverserange(1,10)
    

10 9 8 7 6 5 4 3 2 1 

In [29]:
#function to print the odd numbers in reverse order in a range
def reverseodd(start,end):
    for i in range(end,start-1,-1):
        if i%2!=0:
            print(i,end=" ")
    return
reverseodd(123,130)

129 127 125 123 

In [30]:
#function to calculate the sum of numbers in a range

def sumrange(start,end):
    sum=0
    for i in range(start,end+1):
        sum=sum+i
    return sum
sumrange(1,10)

55

In [35]:
#function to calculate the average of a given range
def avgofrange(start,end):
    sum=0
    count=0
    for i in range(start,end+1):
        count=count+1
        sum=sum+i
        a=sum/count
    print(a)
    return
avgofrange(1,10)
        

5.5


In [ ]:
#function to calculate number of days for a given period
#function to calculate number of hours for a given period

def numberofdays(startyear,endyear):
    sum=0
    for year in range(startyear,endyear+1):
        




## Advanced Problem Set( Optional )
### Function to calculate average of all factorials in a given range
### Function to generate N odd armstrong numbers 
### Function to generate Multiplication table for a number in a given range
#### 10 in the range(100, 102) inclusive
#### 10 x 100 = 1000
#### 10 x 101 = 1010
#### 10 x 102 = 1020

In [42]:
#Function to calculate average of all factorials in a given range
def factavg():
   s=1
   sum=0
   l=int(input('enter 1st no  : '))
   u=int(input('enter 2nd no  : '))
   for i in range(1,u+1):
       s=s*i
       #print(s)
       sum=sum+s
   return sum/u
factavg()


enter 1st no  : 1
enter 2nd no  : 5


30.6